[APM] Agent Configuration bug fixes & enhancements

ogupte opened the pull request:
- added api validation for creating/updating configurations
  - store `transaction_sample_rate` as a number rather than a string
  - improved ui validation for `transaction_sample_rate` precision
  - fixed inconsistent number input from localized input
  - fixed dropdown initial empty selection cross browser issue
  - renamed `settings.sample_rate` to `settings.transaction_sample_rate`
  - fixed yellow cluster state by setting `index.auto_expand_replicas` to `'0-1'`
  - create `.apm-agent-configuration` index on settings page load if it doesn't already exist

![agent-remote-config-ui-4](https://user-images.githubusercontent.com/1967266/61170236-e6b70e80-a51a-11e9-9c1d-6af263245b94.gif)

ogupte committed:
[APM] bug fixes for agent configuration:
  - fix inconsistent number input from localized input
  - fix dropdown initial empty selection cross browser issue
  - rename settings.sample_rate to settings.transaction_sample_rate
  - set 'index.auto_expand_replicas': '0-1'
  - create .apm-agent-configuration index on settings page load if it doesn't already exist

ogupte committed:
removed commented code

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7281/)

ogupte committed:
[APM] add api validation for creating/updating configurations
- store transaction_sample_rate as a number rather than a string
- improved ui validation for transaction_sample_rate precision

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7287/)

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyoutBody.tsx:
```diff
@@ -195,28 +228,26 @@ export function AddSettingFlyoutBody({
             !(
               (Boolean(selectedConfig) &&
-                (isNaN(sampleRate) || isSampleRateValid)) ||
+                (isNaN(sampleRateFloat) || isSampleRateValid)) ||
```
Just wondering, can we not share the Joi schema that we use in the API?

ogupte replied:
We don't currently use Joi on the client side. Joi isn't an isomorphic package as it relies heavily on native node modules. We could try to leverage something like `joi-browser` if we wanted to share schemas.

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyoutBody.tsx:
```diff
@@ -10,17 +10,49 @@ import {
 import { ENVIRONMENT_NOT_DEFINED } from '../../../../../common/environment_filter_values';
 
+const NO_SELECTION = 'NO_SELECTION';
+const EuiSelectWithPlaceholder: typeof EuiSelect = props => (
```
Maybe a comment here that it's a workaround for the EUI/Chrome bug?

sqren replied:
+1. Comment and extract it to a separate file.

dgieselaar reviewed (approved):
couple of questions, otherwise LGTM, slick improvement 👍

sqren commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx:
```diff
@@ -82,8 +84,11 @@ export function AddSettingsFlyout({
+  const isSampleRateValid =
+    sampleRateFloat >= 0 &&
+    sampleRateFloat <= 1 &&
+    sampleRateFloat - parseFloat(sampleRateFloat.toFixed(3)) === 0;
```
Is this to ensure a maximum of three decimals? It's probably a good idea to extract it and give it a sensible name.

```js
const correctDecimals = Number.isInteger(sampleRateFloat * 1000)
```

Or if we want to make it even more clear:
```js
function countDecimals(value: string) {
  if (!value.includes('.')) {
    return 0;
  }
  return value.split('.')[1].length;
}
```

```js
const correctDecimals = countDecimals() < 4
```

sqren commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyoutBody.tsx:
```diff
@@ -43,8 +76,9 @@ export function AddSettingFlyoutBody({
-  sampleRate: number;
-  setSampleRate: React.Dispatch<React.SetStateAction<number>>;
+  sampleRate?: string;
+  setSampleRate: React.Dispatch<React.SetStateAction<string | undefined>>;
```
Are there any advantages to using `React.Dispatch` instead of this (which is a bit easier to read):

```js
setEnvironment: (env: string | undefined) => void;
```

sqren commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx:
```diff
@@ -59,8 +59,10 @@ export function AddSettingsFlyout({
+  const [sampleRate, setSampleRate] = useState<string | undefined>(
+    selectedConfig
+      ? selectedConfig.settings.transaction_sample_rate.toString()
+      : undefined
```
If you change this to:
```js
  const [sampleRate, setSampleRate] = useState(
    selectedConfig
      ? selectedConfig.settings.transaction_sample_rate.toString()
      : ''
```

won't that solve some issues downstream where you have to check whether `sampleRate` is defined and then default to `''`?

sqren commented on x-pack/legacy/plugins/apm/public/components/app/Settings/AddSettings/AddSettingFlyout.tsx:
```diff
@@ -143,6 +148,7 @@ export function AddSettingsFlyout({
             serviceName={serviceName}
             setServiceName={setServiceName}
             sampleRate={sampleRate}
+            sampleRateFloat={sampleRateFloat}
```
Is `sampleRateFloat` really needed in `AddSettingFlyoutBody`? I can see we use it for some validation but `isSampleRateValid` is already passed as an argument. Why is that not sufficient?

sqren replied:
I think if possible `AddSettingFlyoutBody` should not have any of this validation logic. Instead it should just receive the values it needs and matching booleans for whether they are valid.

sqren replied:
Or they could be grouped in pairs:

```js
const sampleRate = {
  value: '12.43',
  isValid: true
}
```

ogupte committed:
create index lazily when the apm-server queries for configs & code cleanup from feedback

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7587/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7688/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7722/)

sorenlouv reviewed (approved):
LGTM! Thanks for getting this in so fast. I think this should be merged now, so we are sure it'll make it out but if you have some spare time on a rainy day, it might make sense to go over these booleans and see if they can be simplified:

```js
const showEnvironmentWarning = !(
    selectedConfig ||
    (!selectedConfig && environment && isSelectedEnvironmentValid && environmentStatus === 'success') ||
    isEmpty(sampleRate)
)

// ...

const showSampleRateWarning = !(
    (Boolean(selectedConfig) && (isEmpty(sampleRate) || isSampleRateValid)) ||
    (!selectedConfig && (!(serviceName || environment) || (isEmpty(sampleRate) || isSampleRateValid)))
)
```

ogupte commented:
> LGTM! Thanks for getting this in so fast. I think this should be merged now, so we are sure it'll make it out but if you have some spare time on a rainy day, it might make sense to go over these booleans and see if they can be simplified:
> 
Totally agree. I really just brain dumped my logic into expanded form bool expressions. and it was easier for me to think in terms of "what is valid" rather than invalid, which is the reason for the root negation, but this could definitely use some simplifying.
